### PR TITLE
Report a validation failure to the PS API service bus

### PR DESCRIPTION
**Pull request by whytheplatypus**
Publishing reports on the validation status of an upload.

work done with:
- @cfarmer-fearless 
- @tedtally 
- @dancing-goats 
- @USZ7 
- @dtisza1

### CODE REVIEW

File: upload-server/cmd/cli/hooks.go [hunk 1]
```diff
@@ -67,6 +69,9 @@ func PrebuiltHooks(appConfig appconfig.AppConfig) (tusHooks.HookHandler, error)
 		Loader: &FileConfigLoader{
 			FileSystem: os.DirFS(appConfig.UploadConfigPath),
 		},
```
Comment (whytheplatypus) [inline upload-server/cmd/cli/hooks.go:71]: can be moved to `internal/loaders/file/file.go` or similar

File: upload-server/cmd/cli/hooks.go [hunk 2]
```diff
@@ -67,6 +69,9 @@ func PrebuiltHooks(appConfig appconfig.AppConfig) (tusHooks.HookHandler, error)
...
+		Reporter: &metadata.FileReporter{
+			Dir: appConfig.LocalReportsFolder,
+		},
```
Comment (whytheplatypus) [inline upload-server/cmd/cli/hooks.go:74]: can be moved to `internal/reporters/file/file.go` or similar

File: upload-server/cmd/cli/hooks.go [hunk 3]
```diff
@@ -78,8 +83,20 @@ func PrebuiltHooks(appConfig appconfig.AppConfig) (tusHooks.HookHandler, error)
...
 			ContainerName: appConfig.AzureManifestConfigContainer,
 		}
+
+		if appConfig.ServiceBusConnectionString != "" {
+
+			sbclient, err := azservicebus.NewClientFromConnectionString(appConfig.ServiceBusConnectionString, nil)
+			if err != nil {
+				return nil, err
+			}
+			preCreateHook.Reporter = &metadata.ServiceBusReporter{
```
Comment (whytheplatypus) [inline upload-server/cmd/cli/hooks.go:93]: rename `preCreateHook` to `manifestValidator` or similar

File: upload-server/internal/metadata/metadata.go [hunk 4]
```diff
@@ -88,47 +92,270 @@ func getVersionFromManifest(ctx context.Context, manifest handler.MetaData, load
...
+}
+
+type Report struct {
+	UploadID        string `json:"upload_id"`
+	StageName       string `json:"stage_name"`
+	DataStreamID    string `json:"data_stream_id"`
+	DataStreamRoute string `json:"data_stream_route"`
+	ContentType     string `json:"content_type"`
+	Content         any    `json:"content"`
+}
```
Comment (whytheplatypus) [inline upload-server/internal/metadata/metadata.go:122]: move to `internal/reports/report.go` or similar

File: upload-server/internal/metadata/metadata.go [hunk 5]
```diff
@@ -88,47 +92,270 @@ func getVersionFromManifest(ctx context.Context, manifest handler.MetaData, load
...
+
+func (v *ValidationError) MarshalJSON() ([]byte, error) {
+	errs := unwrap(v.Err)
+	res := make([]any, len(errs))
+	for i, e := range errs {
+		res[i] = e.Error() // Fallback to the error string
+	}
+	return json.Marshal(res)
+}
```
Comment (whytheplatypus) [inline upload-server/internal/metadata/metadata.go:170]: move to `internal/metadata/validation/errors.go`

File: upload-server/internal/metadata/metadata.go [hunk 6]
```diff
@@ -88,47 +92,270 @@ func getVersionFromManifest(ctx context.Context, manifest handler.MetaData, load
...
+                        "schema_version": "1.0",
+                        "tguid": tguid,
+                        "offset": offset,
+                        "size": size,
+                        "filename": filename,
+                        "data_stream_id": metadata["data_stream_id"],
+                        "data_stream_route": metadata["data_stream_route"],
+                        "metadata": metadata
+            },
+*/
```
Comment (whytheplatypus) [inline upload-server/internal/metadata/metadata.go:196]: remove

File: upload-server/internal/metadata/metadata.go [hunk 7]
```diff
@@ -88,47 +92,270 @@ func getVersionFromManifest(ctx context.Context, manifest handler.MetaData, load
...
+	return hex.EncodeToString(id)
+}
+
+type Identifiable interface {
+	Identifier() string
+}
+
+type Reporter interface {
+	Publish(context.Context, Identifiable) error
+}
```
Comment (whytheplatypus) [inline upload-server/internal/metadata/metadata.go:251]: move to `internal/reports/report.go` or similar

File: upload-server/internal/metadata/metadata.go [hunk 8]
```diff
@@ -88,47 +92,270 @@ func getVersionFromManifest(ctx context.Context, manifest handler.MetaData, load
...
+		}
+	}
+	f, err := os.Create(fr.Dir + "/" + r.Identifier())
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	encoder := json.NewEncoder(f)
+	return encoder.Encode(r)
+}
```
Comment (whytheplatypus) [inline upload-server/internal/metadata/metadata.go:271]: move to `internal/reports/file/reporter.go` or similar

File: upload-server/internal/metadata/metadata.go [hunk 9]
```diff
@@ -88,47 +92,270 @@ func getVersionFromManifest(ctx context.Context, manifest handler.MetaData, load
...
+	if err != nil {
+		return err
+	}
+
+	m := &azservicebus.Message{
+		Body: b,
+	}
+
+	return sender.SendMessage(ctx, m, nil)
 }
```
Comment (whytheplatypus) [inline upload-server/internal/metadata/metadata.go:298]: move to `internal/reports/servicebus/reporter.go` or similar

File: upload-server/pkg/azureinspector/azureinspector.go
Comment (whytheplatypus) [inline upload-server/pkg/azureinspector/azureinspector.go:1]: not sure why it's marking this as a diff, it doesn't even look formatted differently.

**Review by whytheplatypus — COMMENTED** [review]
notes on potential modules

**Comment by whytheplatypus** [comment]
This covers
- UPLOAD-1185
- UPLOAD-1105

File: upload-server/cmd/cli/hooks.go (hunk 3 above)
Comment (dancing-goats) [reply upload-server/cmd/cli/hooks.go:116]: 100% this.

File: upload-server/testing/testing.go [hunk 10]
```diff
@@ -122,84 +122,82 @@ var Cases = map[string]testCase{
...
 
 	if err != nil {
-		return fmt.Errorf("failed to open test file %w", err)
+		return "", fmt.Errorf("failed to open test file %w", err)
 	}
 
 	defer f.Close()
-	paths := []string{"/files/", "/files"}
 
-	for _, path := range paths {
```
Comment (whytheplatypus) [inline upload-server/testing/testing.go:135]: please put the testing back that was removed in the merge

Comment (cfarmer-fearless) [reply upload-server/testing/testing.go:135]: @dancing-goats ^

Comment (dancing-goats) [reply upload-server/testing/testing.go:135]: Where is `paths` used? I checked all the previous commits, and I don't see this being added.

Comment (whytheplatypus) [reply upload-server/testing/testing.go:135]: it's in main, it dropped the for loop when during a merge

File: upload-server/cmd/cli/hooks.go (hunk 1 above)
Comment (tedtally) [reply upload-server/cmd/cli/hooks.go]: created new folder & file paths: 
/internal/loaders/file/file.go
/internal/loaders/azure/azure.go

File: upload-server/cmd/cli/hooks.go (hunk 2 above)
Comment (tedtally) [reply upload-server/cmd/cli/hooks.go]: new folders and files were created and integrated imports into metadata.go

File: upload-server/cmd/cli/hooks.go (hunk 3 above)
Comment (tedtally) [reply upload-server/cmd/cli/hooks.go]: done.

File: upload-server/internal/metadata/metadata.go (hunk 5 above)
Comment (tedtally) [reply upload-server/internal/metadata/metadata.go:181]: selected code was migrated to the errors.go file

File: upload-server/internal/metadata/metadata.go (hunk 7 above)
Comment (tedtally) [reply upload-server/internal/metadata/metadata.go]: Created internal/reporters/reporters.go , migrated the code and integrated the imports into metadata.go

File: upload-server/internal/metadata/metadata.go (hunk 8 above)
Comment (tedtally) [reply upload-server/internal/metadata/metadata.go]: Created internal/reporters/file/file.go , migrated the code and integrated the imports into metadata.go

File: upload-server/internal/metadata/metadata.go (hunk 6 above)
Comment (tedtally) [reply upload-server/internal/metadata/metadata.go:207]: removed

File: upload-server/internal/metadata/metadata.go [hunk 11]
```diff
@@ -146,3 +272,52 @@ func WithTimestamp(event handler.HookEvent, resp hooks.HookResponse) (hooks.Hook
...
+		UploadID:        tguid,
+		DataStreamID:    getDataStreamID(manifest),
+		DataStreamRoute: getDataStreamRoute(manifest),
+		StageName:       "dex-upload-status",
+		ContentType:     "json",
+		DispositionType: "replace",
+		Content:         content,
+	}
+
+	defer func() {
```
Comment (whytheplatypus) [inline upload-server/internal/metadata/metadata.go:301]: defer isn't necessary here since it's the last action.

Comment (tedtally) [reply upload-server/internal/metadata/metadata.go:301]: done.

**Pull request merged**